File: result.py
from dataclasses import dataclass
# 
from typing import Generic, TypeVar

T = TypeVar("T")

@dataclass
class Error:
    message: str
    code: str | None = None
    details: dict | None = None


@dataclass
class Result(Generic[T]):
    """Generic success/error wrapper"""
    value: T | None = None
    error: Error | None = None

    @property
    def is_success(self) -> bool:
        return self.error is None

    @property
    def is_error(self) -> bool:
        return self.error is not None


# Optional: helper functions to make creation cleaner
def Ok(value: T) -> Result[T]:
    return Result(value=value)


def Err(message: str, code: str | None = None, details: dict | None = None) -> Result:
    return Result(error=Error(message, code, details))

File: src/types/result.ts
export type Result<T, E = string> =
| { ok: true; value: T }
| { ok: false; error: E };

export const Ok = <T>(value: T): Result<T, never> => ({ ok: true, value });
export const Err = <E = string>(error: E): Result<never, E> => ({ ok: false, error });

export function getOrElse<T>(result: Result<T, any>, defaultValue: T): T {
    return result.ok ? result.value : defaultValue;
}
